server.go

```diff
@@ -480,29 +442,94 @@ func (a *App) reply(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	// locking?
-	_, err = a.redis.LPush(a.ctx, "msgbus.system.reply", body).Result()
+	_, err = a.redis.LPush(a.ctx, "msgbus.system.reply", body)
 	if err != nil {
 		err = errors.Wrap(err, "couldn't push entry to reply queue")
 		w.Write(errorReply(err.Error()))
 	}
 	w.Write(successReply())
 
 }
+func (a *App) Serve(ctx context.Context) error {
+	a.ctx = ctx
+	go a.runServer(ctx)
+	var serverErr error
+	go func() {
+		if serverErr = a.server.ListenAndServe(); serverErr != http.ErrServerClosed {
+			log.Error().Err(serverErr).Msg("server ListenAndServe failed")
+		}
+	}()
+	interval := 2 * time.Second
+	timer := time.NewTimer(interval)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug().Msg("shutting down the server")
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+			err := a.Cancel(ctx)
+			if err != nil {
+				log.Err(err)
+			}
+			if serverErr != nil && serverErr != http.ErrServerClosed {
+				return errors.Wrap(serverErr, "server ListenAndServe failed")
+			} else if err != nil {
+				return errors.Wrap(err, "error while stopping server")
+			}
+			return err
+		case <-timer.C:
+			timer.Reset(interval)
+		}
+	}
+}
+
+func (a *App) Cancel(ctx context.Context) error {
```

Comment:
This should not be "public" because the user of this server can (must) stop the server by canceling the ctx passed to `Server`. calling cancel is a problem because this will only shutdown the http interface ! not the entire server.

You have to think of rmb as one service that must run as a single process, and can't be separately started or shutdown